# dkfunctions

commonly used bioinformatic functions

- peak_overlap_MC: Monte Carlo simulation of peak overlaps - liklihood over emperical random background
- gsea_dotplot: summary dotplot of multiple gsea comparisons (such as time-courses)
- overlap_two: overlap two peak sets, generate venn, annotate unique and common peaks
- overlap_three
- splic_bar: generate bar graph of splicing changes
- plot_col: genearate violin/swarmplot/boxplot with ttest
- scatter_regression
- signature_heatmap
- ssh_job: generate remote job execution for LSF
- ssh_check: monitors remote jobs
- deeptools: generate deeptools scripts for remote exectution
- order_cluster: supervised clustered heatmap
- ranked_ordered_cluster:  supervised clustering and variable ranking heatmaps
- enrichr: implementation of the enrichr API with barplot functionality
- gsea_barplot: single gsea run summary barplot
- hinton: hinton diagram generation
- genomic_annotation_plots: generate annotation bar graphs
- extract_ENCODE_report_data: extract and plot qc metrics from ENCODE chipseq pipeline
- meme_ssh: submit meme job to LSF
- ovelrap_four: four way overlap of bed files
- extract_clustermap_clusters: return indicies of clustered genes based on clustergram 
- generate_ROSE_gffs: for superenahcner determination
- active_enhancer_determination: call putative active enhancers based on signal and peak files
- boxplot_significance: boxplot with significance testing for ttest, ks test, MW-U
- boxplot_significance_hue: add hue to boxplots
- logrank_km_hazard_ratio: determine the Hazard Ratio of a KaplanMeier plot between two series in lifelines censor format
- scale_factor: active motif format chipseq spike in quantitaiton
- spike_in_plot: plot of chipseq spike in post scale_factor norm
- ENCODE_clean: clean chipseq_encode pipeline folder
- normal_equation: analytically solve linear regresision with normal equation
- chouchou/bliss plots: synergy
- annotate_peaks
- plot_venn2
- plot_venn2_set
- plot_venn3_set
- plot_venn3_counts
